pkg/controller/nodekeeper/nodekeeper_controller.go

```diff
@@ -0,0 +1,169 @@
+	// Create a new controller
+	c, err := controller.New("nodekeeper-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource Node, status change will not trigger a reconcile
+	err = c.Watch(&source.Kind{Type: &corev1.Node{}}, &handler.EnqueueRequestForObject{}, StatusChangedPredicate)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReconcileNodeKeeper implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileNodeKeeper{}
+
+// ReconcileNodeKeeper reconciles a NodeKeeper object
+type ReconcileNodeKeeper struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client               client.Client
+	scheme               *runtime.Scheme
+	metricsClientBuilder metrics.MetricsBuilder
+	machinery            machinery.Machinery
+}
+
+// Reconcile reads that state of the cluster for a UpgradeConfig object and makes changes based on the state read
+// and what is in the UpgradeConfig.Spec
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileNodeKeeper) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling NodeKeeper")
+
+	// Determine if the cluster is upgrading
+	yes, err := r.machinery.IsUpgrading(r.client, "worker", reqLogger)
+	if err != nil {
+		// An error occurred, return it.
+		return reconcile.Result{}, err
+	} else if !yes {
+		// Nodes are not upgrading.
+		return reconcile.Result{}, nil
+	}
+
+	reqLogger.Info("Cluster is upgrading. Proceeding.")
+
+	// Initialise metrics
+	metricsClient, err := r.metricsClientBuilder.NewClient(r.client)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	pdbForceDrainTimeout, err = getPDBForceDrainTimeout(r.client, reqLogger)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			reqLogger.Info("UpgradeConfig not found. No further action.")
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	reqLogger.Info("Checking for PodDisruptionBudget alerts.")
+
+	found, err := checkPDBAlerts(metricsClient)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if !found {
+		log.Info("Found no PDB alerts. No further action")
+		return reconcile.Result{}, nil
+	}
+
+	log.Info(fmt.Sprintf("Found PodDisruptionBudgetAtLimit Alert, Force Drain at: %d minutes", pdbForceDrainTimeout))
+
+	return reconcile.Result{}, nil
+
+}
+
+func checkPDBAlerts(mC metrics.Metrics) (bool, error) {
+	pdbAlertName := "PodDisruptionBudgetAtLimit"
+	var pdbAlerts *metrics.AlertResponse
+	pdbAlerts, err := mC.Query(fmt.Sprintf("ALERTS{alertname=\"%s\"}", pdbAlertName))
```

Comment:
We moved from querying prometheus to getting the PDB directly. Lets do that here too.
Check the `nodeDrained` func in osd upgrader